### ISSUE TITLE
WSGI Adapter does not seem to release resources (spec inconformance)

### ISSUE THREAD

**Issue by apollo13** [pgjones/hypercorn#154]
This is a follow-up to a fediverse discussion with @pgjones. I am not using hypercorn myself, so if I missed something in the codebase I am sorry for the extra noise this issue causes.

Looking at:
https://github.com/pgjones/hypercorn/blob/19dfb96411575a6a647cdea63fa581b48ebb9180/src/hypercorn/app_wrappers.py#L104

hypercorn directly consumes the iterable returned by the app. PEP 333 and 3333 (the quote is from the latter) state:
> If the iterable returned by the application has a close() method, the server or gateway must call that method upon completion of the current request, whether the request was completed normally, or terminated early due to an application error during iteration or an early disconnect of the browser. (The close() method requirement is to support resource release by the application.

As far as I can tell hypercorn should be using something along the lines of:
```python
        response_body_iter = self.app(environ, start_response)
        try:
            for output in response_body_iter:
                if output: # No point in sending empty data
                    send({"type": "http.response.body", "body": output, "more_body": True})
            send({"type": "http.response.body", "body": b"", "more_body": False}) # Note that this is also sent in line 83 already and should probably move up into the `InvalidPathError` there so we don't send it twice.
       finally:
          if hasattr(response_body_iter, 'close'):
               response_body_iter.close()
```

Going even further I think the headers should not be sent in `start_response` directly but only once the first item of the iterable yields results. This would allow the server to properly send back a 500 error when something happens between the call to start_response and the construction of the response iterable as otherwise you already have sent a status code 200 etc…

Please don't hesitate to ask if anything is unclear in my comment and thank you for your work on hypercorn!

**Comment by apollo13** [comment]
I have pushed a potential fix to #155 (without tests), but please note that I do not know the codebase well… 

Btw, https://hypercorn.readthedocs.io/en/latest/how_to_guides/wsgi_apps.html says:
> The full response from the WSGI app will be stored in memory before being sent.

From what I can tell (I am yielding from a wsgi app and adding a sleep in between the yields) the above statement is not true and the response is streamed.

**Comment by pgjones** [comment]
Fixed in 2d2c62bac7b83a8c6766fe3a517f63ff842e5c38
